### PR TITLE
Number input dev, test cases will be added later, Will merge after formatter api is released in cloud sdk api[WIP]

### PR DESCRIPTION
**Pull request by deepprakash345**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes and the overall coverage did not decrease.
- [ ] All unit tests pass on CircleCi.
- [ ] I ran all tests locally and they pass.

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/FormMetaDataDataSourceServlet.java [hunk 1]
```diff
@@ -16,13 +16,7 @@
...
 
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Set;
-import java.util.Spliterator;
-import java.util.Spliterators;
+import java.util.*;
```
Comment (nit23uec) [inline bundles/core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/FormMetaDataDataSourceServlet.java:19]: avoid wild card imports.

File: bundles/core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/FormMetaDataDataSourceServlet.java [hunk 2]
```diff
@@ -122,18 +122,51 @@ protected void doGet(@NotNull SlingHttpServletRequest request, @NotNull SlingHtt
...
+                                String key = entry.getKey();
+                                // This condition is sufficient for property guideComponentType to get ignored
+                                if (key.startsWith("pattern")) {
+                                    Map<String, Object> respObj = new HashMap<>();
+                                    String value = (String) entry.getValue();
+                                    String[] arr = value.split("=", 2);
+                                    respObj.put("text", arr[0]);
+                                    respObj.put("value", arr[1]);
+                                    ValueMap vm = new ValueMapDecorator(respObj);
+                                    resources.add(new ValueMapResource(resourceResolver, path, "nt:unstructured", vm));
```
Comment (nit23uec) [inline bundles/core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/FormMetaDataDataSourceServlet.java:159]: use `JcrConstants` for n`t:unstructured` constant

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/README.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+The component provides a `core.forms.components.textinput.v1` client library category that contains a JavaScript
+component. It should be added to a relevant site client library using the `embed` property.
+
+It also provides a `core.forms.components.textinput.v1.editor` editor client library category that includes
+JavaScript handling for dialog interaction. It is already included by its edit dialog.
+
+## BEM Description
+```
+BLOCK cmp-adaptiveform-textinput
+    ELEMENT cmp-adaptiveform-textinput__label
```
Comment (nit23uec) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/README.md:53]: pls change this for number inout accordingly.

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/_cq_dialog/.content.xml [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                                    value="true"
+                                    name="./exclusiveMaximum"/>
+                        </items>
+                    </basic>
+                    <patterns
+                            jcr:primaryType="nt:unstructured"
+                            jcr:title="Patterns"
+                            sling:resourceType="granite/ui/components/coral/foundation/container">
+                        <items jcr:primaryType="nt:unstructured">
+<!--                            <displayPatternGroup-->
```
Comment (nit23uec) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/_cq_dialog/.content.xml:125]: pls remove commented part

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/numberinput.html [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            type="number"
+            name="${numberinput.name}"
+            value="${numberinput.default}"
+            max="${numberinput.maximum}"
+            min="${numberinput.minimum}"
+            readonly="${numberinput.readOnly}"
+            required="${numberinput.required}"
+            data-cmp-hook-adaptiveform-numberinput="input"
+            placeholder="${numberinput.placeHolder}"
+            data-cmp-required-message="${numberinput.constraintMessages.required}"/>
```
Comment (nit23uec) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/numberinput.html:49]: no need of required message attribute.

File: ui.tests/test-module/specs/numberinput/authoring.spec.js [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            cy.get("[name='./exclusiveMaximum']").should("exist");
+            cy.get('.cq-dialog-cancel').click({force:true});
+            cy.deleteComponentByPath(numberInputDrop) ;
+        });
+
+        it(' type dropdown in Basic tab in edit dialog of NumberInput',function () {
+            dropNumberInputInContainer();
+            cy.openEditableToolbar(sitesSelectors.overlays.overlay.component + numberInputEditPathSelector);
+            cy.invokeEditableAction(editDialogConfigurationSelector);
+            cy.get('.cmp-adaptiveform-numberinput__editdialog').contains('Validation').click({force:true});
```
Comment (nit23uec) [inline ui.tests/test-module/specs/numberinput/authoring.spec.js:110]: better to create  constants for bem. spl for the block.

**Comment by deepprakash345** [comment]
@nit23uec , incorporated review comments

File: bundles/core/pom.xml [hunk 7]
```diff
@@ -456,9 +456,21 @@
 
         <!-- AEM and forms sdk versions -->
         <!-- the below order of dependencies is very critical -->
+<!--        <dependency>-->
+<!--            <groupId>com.adobe.aemds.guide</groupId>-->
+<!--            <artifactId>aemds-guide-core</artifactId>-->
+<!--            <version>7.0.727-SNAPSHOT</version>-->
+<!--        </dependency>-->
```
Comment (vdua) [inline bundles/core/pom.xml:463]: can you please remove that.

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/README.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    ELEMENT cmp-adaptiveform-numberinput__longdescription
+```
+
+## JavaScript Data Attribute Bindings
+
+Apply a `data-cmp-is="adaptiveFormNumberInput"` attribute to the wrapper block to enable initialization of the JavaScript component.
+
+The following attributes can be added to the same element to provide options:
+
+1. `data-cmp-required-message` - populated with requiredMessage from the component's edit dialog
```
Comment (vdua) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/README.md:71]: need to remove this.

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/numberinput.html [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+            class="cmp-adaptiveform-numberinput__widget"
+            aria-label="${numberinput.label.visible && numberinput.label.value}"
+            type="number"
+            name="${numberinput.name}"
+            value="${numberinput.default}"
+            max="${numberinput.maximum}"
+            min="${numberinput.minimum}"
+            readonly="${numberinput.readOnly}"
+            required="${numberinput.required}"
+            data-cmp-hook-adaptiveform-numberinput="input"
```
Comment (vdua) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/numberinput.html:47]: this hook is not needed

**Review by vdua — CHANGES_REQUESTED** [review]

File: bundles/core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/FormMetaDataDataSourceServlet.java [hunk 10]
```diff
@@ -29,11 +31,14 @@
 import javax.servlet.Servlet;
 import javax.servlet.ServletException;
 
+import com.day.crx.JcrConstants;
```
Comment (rismehta) [inline bundles/core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/FormMetaDataDataSourceServlet.java:34]: JcrConstants is incorrectly imported

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/README.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+A `data-cmp-hook-adaptiveform-numberinput="input"` attribute should be added to the input field or textarea so that the JavaScript is able to target it.
```
Comment (rismehta) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/README.md:73]: This needs to be removed

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/clientlibs/site/.content.xml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<jcr:root xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
+    jcr:primaryType="cq:ClientLibraryFolder"
+    categories="[core.forms.components.numberinput.v1.runtime]"
+    dependencies="[core.forms.components.runtime]"/>
```
Comment (vdua) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/clientlibs/site/.content.xml:5]: dependencies should be updated

File: bundles/core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/FormMetaDataDataSourceServlet.java [hunk 13]
```diff
@@ -122,18 +131,51 @@ protected void doGet(@NotNull SlingHttpServletRequest request, @NotNull SlingHtt
...
+                case FORMATTERS:
+                    metaDataList = formMetaData.getFormatters(fieldType);
+                    while (metaDataList.hasNext()) {
+                        FormsManager.ComponentDescription componentDescription = metaDataList.next();
+                        Resource formatterResource = resourceResolver.getResource(componentDescription.getResourceType());
+                        if (formatterResource != null) {
+                            String path = formatterResource.getPath();
+                            ValueMap formatterResourceValueMap = formatterResource.getValueMap();
+                            Iterator<Map.Entry<String, Object>> it = formatterResourceValueMap.entrySet().iterator();
+                            while (it.hasNext()) {
```
Comment (rismehta) [inline bundles/core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/FormMetaDataDataSourceServlet.java:157]: Please use streams

**Review by rismehta — CHANGES_REQUESTED** [review]
-1 for now

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/clientlibs/site/.content.xml (hunk 12 above)
Comment (nit23uec) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/clientlibs/site/.content.xml:5]: deps needs to be correct.  core.forms.components.runtime.base, container

**Comment by rismehta** [comment]
Can you comment formatters code so that the build passes

File: bundles/core/pom.xml (hunk 7 above)
Comment (deepprakash345) [reply bundles/core/pom.xml:463]: Removed the same

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/README.md (hunk 8 above)
Comment (deepprakash345) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/README.md:71]: Removed the same

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/numberinput.html (hunk 9 above)
Comment (deepprakash345) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/numberinput.html:47]: Removed the same

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/clientlibs/site/.content.xml (hunk 12 above)
Comment (deepprakash345) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/clientlibs/site/.content.xml:5]: updated

File: bundles/core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/FormMetaDataDataSourceServlet.java [hunk 14]
```diff
@@ -122,18 +131,51 @@ protected void doGet(@NotNull SlingHttpServletRequest request, @NotNull SlingHtt
...
         FormMetaData formMetaData = resourceResolver.adaptTo(FormMetaData.class);
         if (formMetaData != null) {
             Iterator<FormsManager.ComponentDescription> metaDataList = null;
             switch (type) {
+                case FORMATTERS:
+                    metaDataList = formMetaData.getFormatters(fieldType);
+                    while (metaDataList.hasNext()) {
+                        FormsManager.ComponentDescription componentDescription = metaDataList.next();
+                        Resource formatterResource = resourceResolver.getResource(componentDescription.getResourceType());
+                        if (formatterResource != null) {
```
[the remark this answers is not in the document]
Comment (deepprakash345) [reply bundles/core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/FormMetaDataDataSourceServlet.java:153]: commented the code as asked the below comment, while merging will change this code to use streams and filters

File: bundles/core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/FormMetaDataDataSourceServlet.java (hunk 10 above)
Comment (deepprakash345) [reply bundles/core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/FormMetaDataDataSourceServlet.java:34]: import org.apache.jackrabbit.JcrConstants;

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/README.md (hunk 11 above)
Comment (deepprakash345) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/README.md:73]: Removed the same

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/clientlibs/site/.content.xml (hunk 12 above)
Comment (deepprakash345) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/clientlibs/site/.content.xml:5]: made the suggested change

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/clientlibs/editor/js/editDialog.js [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+        submitButton.addEventListener("click",_manageDialogSubmit);
+        function _manageDialogSubmit(){
+            var leadDigitsElement = dialog.find(NUMERICINPUT_LEADDIGITS)[0];
+            var fracDigitsElement = dialog.find(NUMERICINPUT_FRACDIGITS)[0];
+            var editFormatElement = dialog.find(NUMERICINPUT_EDITFORMAT)[0];
+            var leadDigits=leadDigitsElement.value;
+            var fracDigits=fracDigitsElement.value;
+            var editFormat="";
+            editFormatElement.value=editFormat;
+            if((leadDigits!=null && leadDigits!="" && leadDigits>0) ||(fracDigits!=null && fracDigits!="" && fracDigits>0)) {
```
Comment (nit23uec) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/clientlibs/editor/js/editDialog.js:63]: can be simplified to `if (!leadDigits || !fragDigits)`

**Review by vdua — APPROVED** [review]

**Comment by vdua** [comment]
@nit23uec @rismehta The build is broken due to a npm dependency. But I do not see any changes in the package.json recently.

File: bundles/core/src/main/java/com/adobe/cq/forms/core/components/models/form/NumberInput.java [hunk 16]
```diff
@@ -28,21 +28,9 @@
...
-
-    /**
-     * Returns the fractional digits for float data type
-     * 
-     * @return fractional digits
-     */
-    @Nullable
-    @JsonIgnore
-    Integer getFracDigits();
+    default String getEditFormat() {
```
Comment (nit23uec) [inline bundles/core/src/main/java/com/adobe/cq/forms/core/components/models/form/NumberInput.java:33]: `editFormat` should not be `JsonIgnore` as it's part of AF2 spec

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/numberinput.html [hunk 17]
```diff
@@ -0,0 +1,53 @@
+<!--/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  ~ Copyright 2021 Adobe
```
Comment (nit23uec) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/numberinput.html:2]: copyright

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/numberinput.js [hunk 18]
```diff
@@ -0,0 +1,33 @@
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ ~ Copyright 2021 Adobe
```
Comment (nit23uec) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/numberinput/v1/numberinput/numberinput.js:2]: copyright

**Pull request merged**